File: JANL/Managers/FormManager.vb
Public Class FormManager
	Private Shared Root As ToolStripMenuItem
	Private Shared TSMIs As New Dictionary(Of Form, ToolStripMenuItem)

	Public Shared Sub Add(Form As Form)
		If TSMIs.ContainsKey(Form) Then Exit Sub
		Dim TSMI As New ToolStripMenuItem With {
			.Text = Form.Text,
			.Image = Form.Icon.ToBitmap()
		}
		AddHandler TSMI.Click, Sub() Form.Activate()
		AddHandler Form.TextChanged, AddressOf TextChanged
		AddHandler Form.FormClosed, AddressOf FormClosed

		Root.DropDownItems.Add(TSMI)
		Root.Visible = True
		TSMIs.Add(Form, TSMI)
	End Sub

	Public Shared Sub SetMenu(TSMI As ToolStripMenuItem)
		If Root IsNot Nothing Then
			For Each KV In TSMIs
				Root.DropDownItems.Remove(KV.Value)
				TSMI.DropDownItems.Add(KV.Value)
			Next
			ChangeVisibility()
		End If
		Root = TSMI
		ChangeVisibility()
	End Sub

	Private Shared Sub FormClosed(sender As Object, e As EventArgs)
		Dim Form = DirectCast(sender, Form)
		Remove(Form)
	End Sub

	Private Shared Sub Remove(Form As Form)
		If Not TSMIs.ContainsKey(Form) Then Exit Sub
		Dim TSMI = TSMIs(Form)
		RemoveHandler Form.TextChanged, AddressOf TextChanged
		RemoveHandler Form.FormClosed, AddressOf FormClosed

		Root.DropDownItems.Remove(TSMI)
		TSMIs.Remove(Form)
		TSMI.Dispose()
		ChangeVisibility()
	End Sub

	Private Shared Sub ChangeVisibility()
		Root.Visible = Not Root.DropDownItems.Count = 0
	End Sub

	Private Shared Sub TextChanged(sender As Object, e As EventArgs)
		Dim Form = DirectCast(sender, Form)
		TSMIs(Form).Text = Form.Text
	End Sub

End Class

File: JANL.WinForms/Managers/FormManager.cs
using System;
using System.Collections.Generic;
using System.Windows.Forms;

namespace JANL.Managers
{
    public static class FormManager
    {
        private static Dictionary<Form, ToolStripMenuItem> Items = new Dictionary<Form, ToolStripMenuItem>();
        private static ToolStripMenuItem Root;

        public static void Add(Form form)
        {
            if (Items.ContainsKey(form)) { return; }
            var TSMI = new ToolStripMenuItem
            {
                Text = form.Text,
                Image = form.Icon.ToBitmap()
            };
            TSMI.Click += (_, __) => { form.Activate(); };
            form.TextChanged += TextChanged;
            form.FormClosed += FormClosed;

            Root.DropDownItems.Add(TSMI);
            Root.Visible = true;
            Items.Add(form, TSMI);
        }

        public static void SetMenu(ToolStripMenuItem TSMI)
        {
            if (Root != null)
            {
                foreach (var KV in Items)
                {
                    Root.DropDownItems.Remove(KV.Value);
                    TSMI.DropDownItems.Add(KV.Value);
                }
                ChangeVisibility();
            }
            Root = TSMI;
            ChangeVisibility();
        }

        private static void ChangeVisibility() => Root.Visible = Root.DropDownItems.Count != 0;

        private static void FormClosed(object sender, FormClosedEventArgs e) => Remove((Form)sender);

        private static void Remove(Form form)
        {
            if (!Items.ContainsKey(form)) { return; }
            var TSMI = Items[form];
            form.TextChanged -= TextChanged;
            form.FormClosed -= FormClosed;

            Root.DropDownItems.Remove(TSMI);
            Items.Remove(form);
            TSMI.Dispose();
            ChangeVisibility();
        }

        private static void TextChanged(object sender, EventArgs e)
        {
            var Form = (Form)sender;
            Items[Form].Text = Form.Text;
        }
    }
}